### ISSUE TITLE
Possible memory leak or ineffective ak.clear()  (not visible to ak.report_mem())

### ISSUE THREAD

**Issue by 21771** [Bears-R-Us/arkouda#2044]
**Describe the bug**
chapel server crashes (SementedString.chpl:280: error: Out of memory allocating "array elements" when loading/processing new data after ak.clear() but allows loading/processing after restarting chapel server and loading the same data. 

ak.util.report_mem() shows same value in both cases, but the server crashes in the former and succeeds in the latter.

**To Reproduce**
General flow:

```
> load stored dataframe
> register dataframe (with name X)
> print(ak.util.report_mem())
> Iterate:
>     load additional data
>     modify dataframe for existing rows
>     append` rows
>     save dataframe
>     unregister dataframe (with name X)
>     register new dataframe (with name X)
>     clear memory (ak.clear())
>     print(ak.util.report_mem())
> 
```

**Expected behavior**
Expect the iterations to fail at some point when dataframe becomes too large. Once it fails, I would not expect subsequent invocations to add data (since it exceeds memory limits). 

However. currently after out-of-memory error and restart, the first report_mem() shows the same value as the one before the previous failed iteration, but the newly restarted chapel server will will successfully complete another iteration before failing.

**Error Message**
If you are receiving any error messages, please provide them here.
"SementedString.chpl:280: error: Out of memory allocating "array elements"

**Is this a Blocking Issue**
- Are there any known work-arounds?
- Restarting server between iterations is impractical, but works.
- 
- If a known work around exists, please provide a code block demonstrating it.

**`ak.get_config()` Output**
Providing the output of `ak.get_config()` provides crucial debugging information.

**Additional context**

**Comment by 21771** [comment]
Simpler model exhibiting same behavior:

```
> print(ak.util.report_mem())
> load dataframe from file(s) X
> Iterate:
>     load additional data from one-up files (next up not already included in dataframe)
>     modify dataframe for existing rows
>     append` rows
>     save dataframe to file(s) X
>     clear memory (ak.clear())
>     print(ak.util.report_mem())
```

In this case out of memory occurs after loading and attempting to process/append data from one-up file N. Restarting will load one-up file N and successfully process/append data and subsequently fail on N+1. After restarting server, N+1 data is successfully added and process fails on process/append of N+2.

**Comment by Ethan-DeBandi99** [comment]
@pierce314159 - I looked into this briefly today. I think we should have a discussion tomorrow.

**Comment by pierce314159** [comment]
@ronawho is there any chance this is coming up because we're not tracking small allocations?

**Comment by ronawho** [comment]
Not tracking small allocations means we don't track aggregator memory usage and arrays-of-heap-allocated values (e.g. arrays of bigints.) Aggregators don't use that much memory and I don't think arrays-of-heap-allocated values would be used here so I would guess not. You could try running the server with `--memThreshold=1` to see if that impacts anything (note that this can significantly degrade performance)

I would guess it's more likely that there's some codepath missing an overmemlimit check.

**Comment by pierce314159** [comment]
> Not tracking small allocations means we don't track aggregator memory usage and arrays-of-heap-allocated values (e.g. arrays of bigints.) Aggregators don't use that much memory and I don't think arrays-of-heap-allocated values would be used here so I would guess not. You could try running the server with `--memThreshold=1` to see if that impacts anything (note that this can significantly degrade performance)
> 
> I would guess it's more likely that there's some codepath missing an overmemlimit check.

okay thanks!!

**Comment by Ethan-DeBandi99** [comment]
@ronawho - have you had a chance to did on this yet? Is there anything we can do on our end?

**Comment by ronawho** [comment]
No, and probably won't get a chance for at least a few days. If you have time to put together a reproducer I could run, that would be really helpful.

**Comment by ronawho** [comment]
Arkouda uses the amount of physical memory as part of the memory estimation checks, but I'm wondering if the amount of allocatable memory is significantly lower than that. The below program reports the amount of physical memory and the amount of memory that can actually be allocated from the chapel runtime. I'd expect them to be within ~5-10% of each other, so if they're not there could be a system issue. And even if they are close, we probably want to use chapel's estimate of available memory instead of using physical memory as a proxy.

```chpl
use Memory.Diagnostics, CTypes;

proc runtimeHeapMem() {
  extern proc chpl_comm_regMemHeapInfo(ref start: c_void_ptr, ref size: c_size_t): void;
  var unused: c_void_ptr;
  var heap_size: c_size_t;
  chpl_comm_regMemHeapInfo(unused, heap_size);
  return heap_size.safeCast(int);
}

writeln("physMem=", here.physicalMemory(unit = MemUnits.Bytes));
writeln("heapMem=", runtimeHeapMem());
```

If you could run the above program on the target system and either report (offline is fine) the raw numbers or just the % difference between them that'd be helpful. I'd also be interested in whether results with `--memThreshold=1` were any different.

**Comment by Ethan-DeBandi99** [comment]
@ronawho - Results of the provided code are below.

**This was run with 10 nodes**

As a separate note, the user reported that their Chapel 1.29 running LLVM would not allow the code to compile. I am waiting on confirmation of what version and backend was used to run the test.

The system is reporting that heap is ~95% of the physical memory.

**Comment by ronawho** [comment]
> The system is reporting that heap is ~95% of the physical memory.

Ok, that seems right. Any change in OOM when running with `--memThreshold=1` or any info on a standalone reproducer?

**Comment by Ethan-DeBandi99** [comment]
This is all I was able to get from them last week. I will try getting more info this week.
